src/HooglePlus/FilterTest.hs

```diff
@@ -0,0 +1,194 @@
+    argName (TyParen _ t) = argName t
+    argName other = throw $ NotSupportedException ("Not able to handle " ++ show other)
+
+    buildRet typeEnv argList (TyForall _ _ (Just ctx) t) = buildRet typeEnv' argList t
+      where typeEnv' = typeEnv ++ buildEnv typeEnv ctx
+    buildRet typeEnv argList (TyFun _ typeArg typeRet) = buildRet typeEnv argList' typeRet
+      where argList' = argList ++ [argName typeArg]
+    buildRet typeEnv argList typeRet = (typeEnv, argList, argName typeRet)
+
+    extractQualified (ClassA _ (UnQual _ (Ident _ name)) var) =
+      map (\x -> (argName x, name)) var
+    extractQualified (ParenA _ qual) = extractQualified qual
+    extractQualified other = throw $ NotSupportedException ("Not able to extract " ++ show other)
+    buildEnv typeEnv (CxSingle _ item) = typeEnv ++ extractQualified item
+    buildEnv typeEnv (CxTuple _ list) = foldr ((++) . extractQualified) typeEnv list
+
+generateRandomValue :: String -> IO String
+generateRandomValue typeName = do
+  result <- runInterpreter $ do {
+    setImports ["Prelude", "Test.QuickCheck"];
+
+    act <- interpret ("do {\
+    \it <- generate (arbitrary :: Gen (" ++ typeName ++ "));\
+    \return $ show it\
+    \}") (as :: IO String);
+
+    liftIO act
+  }
+  case result of
+    Left err -> putStrLn (displayException err) >> return "error \"114514\""
+    Right res -> return ("(" ++ res ++ ")")
+
+mapRandomMultiple :: (TypeEnv, [ArgumentType], ArgumentType) -> Int -> IO [String]
+mapRandomMultiple env count = do
+  base <- mapRandomArguments' env
+  replicateM count (derive base)
+  where
+    derive :: [(String, String)] -> IO String
+    derive base = do
+      rand <- generate (arbitrary :: Gen Int)
+      let index = rand `rem` length base
+      let item = getNthElement index base
+
+      replaced <- newRandomValue item
+      let list = setNthElement index replaced base
+      return $ unwords (map snd list)
+
+    newRandomValue (name, _) = do
+      val <- generateRandomValue name
+      return (name, val)
+
+    getNthElement idx list = item
+      where (_, item:_) = splitAt idx list
+
+    setNthElement idx item list
+      | idx < 0   = list
+      | otherwise = case splitAt idx list of
+                    (front, _:back) -> front ++ item : back
+                    _ -> list
+
+mapRandomArguments :: (TypeEnv, [ArgumentType], ArgumentType) -> IO String
+mapRandomArguments env = do
+  result <- mapRandomArguments' env
+  return $ unwords (map snd result)
+
+mapRandomArguments' :: (TypeEnv, [ArgumentType], ArgumentType) -> IO [(String, String)]
+mapRandomArguments' (typeEnv, args, retType) = mapM f (transformType typeEnv args)
+  where
+    f x = let str = show x in do
+      val <- generateRandomValue str
+      return (str, val)
+
+    transformType [] args = args
+    transformType (envItem:env) args = transformType env args'
+      where
+        args' = map (replace envItem) args
+
+        replace (Polymorphic name, cons) (Polymorphic name') =
```

Comment:
note for the future: `cons` is generally a bad name for a variable, it overloads a function name that's in people's minds when they read this. Something like `rest` or `restX` where `X` is a description of what the rest are of.

Similarly, `x` and `x'` don't really help so much instead of `oldX` `newX`. Generally it should be possible to determine what's going on with each line without having to look up name bindings.